amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseCleanupManager.java

```diff
@@ -0,0 +1,192 @@
+import software.amazon.awssdk.services.kinesis.model.GetShardIteratorResponse;
+import software.amazon.awssdk.services.kinesis.model.ResourceNotFoundException;
+import software.amazon.awssdk.services.kinesis.model.ShardIteratorType;
+import software.amazon.kinesis.common.FutureUtils;
+import software.amazon.kinesis.common.KinesisRequestsBuilder;
+import software.amazon.kinesis.common.StreamIdentifier;
+import software.amazon.kinesis.leases.exceptions.DependencyException;
+import software.amazon.kinesis.leases.exceptions.InvalidStateException;
+import software.amazon.kinesis.leases.exceptions.ProvisionedThroughputException;
+import software.amazon.kinesis.metrics.MetricsFactory;
+import software.amazon.kinesis.retrieval.AWSExceptionManager;
+import software.amazon.kinesis.retrieval.kpl.ExtendedSequenceNumber;
+
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+/**
+ * Helper class to cleanup of any expired/closed shard leases. It will cleanup leases periodically as defined by
+ * {@link LeaseManagementConfig#leaseCleanupIntervalMillis()} upon worker shutdown, following a re-shard event or
+ * a shard expiring from the service.
+ */
+@Slf4j
+@RequiredArgsConstructor
+@EqualsAndHashCode
+public class LeaseCleanupManager {
+    @NonNull
+    private final StreamIdentifier streamIdentifier;
+    @NonNull
+    private final LeaseCoordinator leaseCoordinator;
+    @NonNull
+    private final KinesisAsyncClient kinesisClient;
+    @NonNull
+    private final MetricsFactory metricsFactory;
+    @NonNull
+    private final Duration maxFutureWait;
+    private final boolean cleanupLeasesUponShardCompletion;
+    private final boolean shouldCleanupLeases;
+    private static final int MAX_RECORDS = 1;
+
+    /**
+     * Cleanup the specified lease.
+     *
+     * @param lease
+     * @param shardInfo
+     * @return true if the lease was deleted, false otherwise.
+     * @throws DependencyException
+     * @throws ProvisionedThroughputException
+     * @throws InvalidStateException
+     * @throws TimeoutException
+     */
+    public synchronized boolean cleanupLease(Lease lease, ShardInfo shardInfo) throws DependencyException, ProvisionedThroughputException, InvalidStateException, TimeoutException {
+        if (lease == null) {
+            log.warn("Cannot cleanup lease for {} as this instance does not currently hold it.", ShardInfo.getLeaseKey(shardInfo));
+            return false;
+        }
+
+        final AWSExceptionManager exceptionManager = createExceptionManager();
+        final List<ChildShard> childShards;
+        boolean cleanedUpLease = false;
+
+        try {
+            try {
+                childShards = getChildShards(shardInfo);
```

Comment:
```
shouldCleanCompletedShards =  cleanupLeasesUponShardCompletion && completionTimer(shard).shouldCleanNow();
shouldCleanGarbageShards = garbageTimer(shard).shouldCleanNow();

if(shouldCleanCompletedShards || shouldCleanGarbageShards) {
 // We are trying to reduce the invocations of below call
 childShards = getChildShards(shardInfo);
if(shouldCleanCompletedShards) {
// cleanup logic
}
}

```

Reply:
This should be covered as part of moving this to a scheduled thread pool